Tl/lt 88456218 cleanup login.js

luketower opened the pull request:
[Pivotal Story](https://www.pivotaltracker.com/story/show/88456218)
- [x] Specs pass

Questions under investigation (Must be addressed before merging. Feel free to comment on PR with suggestions/answers):
- What is the best way to test this to make sure all functionality is still present?
- What other apps use this library that need to be modified?
- Is a version bump necessary for this change?

luketower committed:
[TL/LT][88456218] Add karma and gulp. Move formatting into own module.

[Finishes #8845621]

luketower committed:
[TL/LT][88456218] Move error handling into own module.

michaelstalker commented on package.json:
```diff
@@ -13,8 +13,41 @@
+    "build": "gulp dist",
+    "livereload": "gulp",
+    "watch": "gulp",
+    "watch:dev": "gulp",
```
Do we need all these scripts? Three just run `gulp`.

michaelstalker commented on src/error_handler.coffee:
```diff
@@ -0,0 +1,48 @@
+      events.trigger('event/' + eventName, error)
+
+
+    _formatError: (key, value) ->
```
What are your thoughts about removing the underscore? It makes more sense for us to treat this as a public function on an `ErrorHandler` object than it did when this function lived on the `Login` object.

michaelstalker commented on src/error_handler.coffee:
```diff
@@ -0,0 +1,48 @@
+          formatted_key = new Formatter(key).sentenceCase()
+          if value then "#{formatted_key} #{value}" else ''
+
+    _clearErrors: ($div) ->
```
What are your thoughts about removing the underscore? It makes more sense for us to treat this as a public function on an `ErrorHandler` object than it did when this function lived on the `Login` object.

michaelstalker commented:
> What other apps use this library that need to be modified?

AG, and maybe m.AG. @c0: Are there others?

> Is a version bump necessary for this change?

If you want other apps to include this change, yes. I don't think it would be bad to wait to include this change in other apps, though. Maybe we could wait until we change the public behavior of login.js.

luketower committed:
[TL/LT][88456218] Remove underscore in function name to indicate public.

- Remove redundant npm scripts.
- Remove unnecessary require of formatter in login file.

c0 commented on src/error_handler.coffee:
```diff
@@ -0,0 +1,44 @@
+      messages = ''
+      if error?
+        $form = @$box.parent().find 'form'
+        $.each @error, (key, value) =>
```
Coffeescript supports looping out of the box:

`for key, value of @error`

c0 commented on src/formatter.coffee:
```diff
@@ -0,0 +1,12 @@
+    constructor: (@string) ->
+
+    sentenceCase: () ->
+      new_string = @string.replace('_', ' ')
```
`#replace` only works on the first occurrence. Should this replace multiple `_`? e.g. `@string.split('_').join(' ')`

Teapane replied:
Good catch. :+1:

Teapane commented:
- Looks like `Ag` and `mAg` both use this library.  
- We can create a symlink and check functionality on both Apps to verify. I can check `m.Ag` with these changes if you want to check `Ag`?  
- Both apps `bower.json` will need to be updated with the new version.  
- And all the things :star2: 
  - Michael Stalker asks, "Did you _really_ mean **all** the things?" :trollface: 
  - Tyler Long says, " :trollface: :trollface: **ALL** the things." 
  - Michael Stalker replies, "Dang. I guess you **did** mean all the things."

Teapane committed:
[TL/LT][88456218] PR comment

Teapane commented:
Once this is merged, we will need to bump the version to `v0.2.29`. Will also require [PR 1](https://github.com/rentpath/ag/pull/2259) and [PR 2](https://github.com/rentpath/m.apartmentguide/pull/513) to be merged.

luketower committed:
[TL/LT][88456218] Use coffeescript iterator. Add more test coverage.

luketower committed:
[TL/LT][88456218] Add all files for main.

luketower commented:
Please do not merge this yet. We're still solving some existing problems.

luketower committed:
[TL/LT][88456218] Do not reference src dir.

luketower commented:
This PR is ready for more review and possible merge.

luketower committed:
[TL/LT][88456218] Version bump.

luketower committed:
Merge branch 'dev' into tl/lt_88456218_cleanup_login.js

Conflicts:
  bower.json
  login.js
  package.json

michaelstalker commented on README.md:
```diff
@@ -1,14 +1,98 @@
-# login.js
+## Getting started for development.
```
:+1: For all the documentation on the README

michaelstalker commented:
I'll give this a :+1: .

The tests on the private functions were mostly already there, but sometime in the future I think it would be good to ask why we're testing these. They would be better off as public functions--either on the `Login` object or on another object--or untested.
